client/actions/show-heartbeat/index.js

```diff
@@ -21,11 +30,25 @@ export default class ShowHeartbeatAction extends Action {
     } = this.recipe.arguments;
 
     const lastShown = await this.getLastShownDate();
+
+    // get the last shown for heartbeats in general
+    const lastHeartbeatShown = await this.getLastHeartbeatDate();
+
+    // we can show the heartbeat if there has not been one shown recently,
+    const canShowHeartbeat = lastHeartbeatShown === null
+    // or if the last one shown was more than HEARTBEAT_THROTTLE ms ago
+      || Date.now() - lastHeartbeatShown >= HEARTBEAT_THROTTLE;
```

Comment:
I think it is a little thing, and it doesn't matter too much. I'd put the comments all together above the entire expression though. Plus, it seems a little odd to pull out `canShowHeartbeat` into its own expression, instead of just having it be a subexpression of `shouldShowSurvey`.

I think that the complexity of this logic is a sign that it could be better abstracted.

Maybe it makes sense to move some more logic into the helper functions. Here I'm using `Infinity` as a trick to represent never, though maybe it would make sense to even move the comparisons into the helpers?

```js
if (!testing && (
  await this.timeSinceLastHeartbeat() < HEARTBEAT_THROTTLE ||
  await this.timeSinceLastSurvey() < LAST_SHOWN_DELAY
)) {
  return;
}

...

/* Calculates the number of milliseconds since the last heartbeat from any recipe
 * was shown. If no heartbeat has ever been shown, returns Infinity, or "forever ago".
 */
async timeSinceLastHeartbeat() {
  const lastShown = await this.heartbeatStorage.getItem('lastShown');
  // If no heartbeat has ever been shown, lastShown will be falsey.
  if (lastShown) {
    return new Date() - lastShown;
  } else {
    return Infinity;
  }
}

/* Calculates the number of milliseconds since the last survey from this recipe
 * was shown. If no survey has ever been shown, returns Infinity, or "forever ago".
 */
async timeSinceLastSurvey() {
  const lastShown = await this.storage.getItem('lastShown');
  // If no survey has been shown, lastShown will be falsey.
  if (lastShown) {
    return new Date() - lastShown;
  } else {
    return Infinity;
  }
}
```